### PR TITLE
Added dynamic `TargetDescription`

### PR DESCRIPTION
**Pull request by theIDinside**
TargetDescription dynamically generates the "master" target description contents (target.xml), so that we only have to manage the individual concrete CPU features (like foo-sse.xml, for instance). TargetDescription will generate the xml contents that includes the used features.

Removed combinatorial target.xml files

These are now handled by `TargetDescription` instead.

This unblocks the #3776 PR (AVX512 support) from getting pulled in, though that PR will need to take this PR into account.

Note: 
The `GdbServerRegister` should probably be rewritten/removed and then rely on a `TargetDescription` that provides a "static" array of `GdbServerValue` since once we know the target description, it won't change during a replay, so dynamically allocating like we do now (in `dispatch_regs_request`) is not really necessary. But these are just loose thoughts I have, I haven't really verified that this would be a better approach.

### CODE REVIEW

**Comment by theIDinside** [comment]
@rocallahan || @khuey 
Does the tests run on actual hardware or some emulation? The errors are confusing; I just can't see how that can possibly be true without there actually existing RR bugs, and that now, with this patch we *actually* provide aarch64 target descriptions and that's why gdb asserts saying it expects 268 bytes (which is the aarch64-core.xml size) but got 788 (which I think is just x86 or some variant, possibly without a few registers)

**Comment by theIDinside** [comment]
Prior to this patch, I have no idea how RR informed GDB on `aarch64` that the `g` packet contains the `aarch64-fpu.xml` because the only file it ever transferred using `xfer` was  `aarch64-core.xml`. But with this patch, I forgot to add the `fpu.xml` in the dynamically generated target description and the tests for aarch64 failed. Now with the latest commit it works. But it does puzzle me why it hasn't failed before (since `fpu.xml` was never actually sent before).

File: src/GdbServerConnection.cc [hunk 1]
```diff
@@ -111,7 +112,8 @@ unique_ptr<GdbServerConnection> GdbServerConnection::await_connection(
     Task* t, ScopedFd& listen_fd, const GdbServerConnection::Features& features) {
   auto dbg = unique_ptr<GdbServerConnection>(
     new GdbServerConnection(t->thread_group()->tguid(), features));
-  dbg->set_cpu_features(get_cpu_features(t->arch()));
+  const auto arch = t->arch();
+  dbg->set_cpu_features(arch, get_cpu_features(arch));
```
Comment (rocallahan) [inline src/GdbServerConnection.cc:116]: You can just drop this hunk.

File: src/GdbServerConnection.cc [hunk 2]
```diff
@@ -609,11 +588,8 @@ bool GdbServerConnection::xfer(const char* name, char* args) {
       return false;
     }
 
-    string target_desc =
-        read_target_desc((strcmp(annex, "") && strcmp(annex, "target.xml"))
-                             ? annex
-                             : target_description_name(cpu_features_));
-    write_xfer_response(target_desc.c_str(), target_desc.size(), offset, len);
+    const auto desc = (strcmp(annex, "") && strcmp(annex, "target.xml") ? read_target_desc(annex) : target_decription->to_xml());
```
Comment (rocallahan) [inline src/GdbServerConnection.cc:591]: Remove outer parentheses here.

File: src/GdbServerConnection.h [hunk 3]
```diff
@@ -877,6 +879,7 @@ class GdbServerConnection {
   bool hwbreak_supported_; // client supports hwbreak extension
   bool swbreak_supported_; // client supports swbreak extension
   bool list_threads_in_stop_reply_; // client requested threads: and thread-pcs: in stop replies
+  std::unique_ptr<TargetDescription> target_decription{nullptr};
```
Comment (rocallahan) [inline src/GdbServerConnection.h:882]: Remove the initializer, it's the same as the default.

File: src/TargetDescription.h [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+#include <cstdint>
+
+namespace rr {
+
+struct GdbServerRegisterValue;
+
+using u32 = std::uint32_t;
+
+enum class TargetFeature : u32 {
+    Core = 0,
```
Comment (rocallahan) [inline src/TargetDescription.h:12]: 2-space indent

File: src/TargetDescription.h [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    SSE,
+    Linux,
+    Segment,
+    AVX,
+    PKeys
+};
+
+class TargetDescription {
+    SupportedArch arch;
```
Comment (rocallahan) [inline src/TargetDescription.h:22]: 2-space indent. Also, put the fields at the end of the class definition

File: src/TargetDescription.h [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    std::vector<TargetFeature> target_features;
+public:
+    explicit TargetDescription(rr::SupportedArch arch, u32 cpu_features) noexcept;
```
Comment (rocallahan) [inline src/TargetDescription.h:25]: `uint32_t`

Comment (rocallahan) [reply src/TargetDescription.h:25]: Also, drop `noexcept`

File: src/TargetDescription.cc [hunk 7]
```diff
@@ -0,0 +1,124 @@
+#include "TargetDescription.h"
+#include "GdbServerConnection.h"
+#include "kernel_abi.h"
+#include <sstream>
+namespace rr {
```
Comment (rocallahan) [inline src/TargetDescription.cc:5]: blank line after last #include

File: src/TargetDescription.cc [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+class FeatureStream {
+  std::stringstream stream{};
```
Comment (rocallahan) [inline src/TargetDescription.cc:8]: private fields go last.
Don't use {}, default initializer is fine.

File: src/TargetDescription.cc [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+  const char* arch_prefix{nullptr};
```
Comment (rocallahan) [inline src/TargetDescription.cc:9]: prefer `= nullptr`.

File: src/TargetDescription.cc [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+public:
+  std::string result() noexcept { return stream.str(); }
+
+  template <typename Any>
+  friend FeatureStream& operator<<(FeatureStream& stream, Any any) noexcept;
```
Comment (rocallahan) [inline src/TargetDescription.cc:15]: Don't use `noexcept` anywhere.

File: src/TargetDescription.cc [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+    case TargetFeature::Segment:
+      stream << "seg.xml";
+      break;
+  }
+  stream << R"("/>)" << '\n';
+  return stream;
+}
+
+TargetDescription::TargetDescription(rr::SupportedArch arch,
+                                     u32 cpu_features) noexcept
```
Comment (rocallahan) [inline src/TargetDescription.cc:80]: uint32_t

File: src/TargetDescription.cc [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+  }
+}
+
+static constexpr auto Header = R"(<?xml version="1.0"?>
+<!DOCTYPE target SYSTEM "gdb-target.dtd">
+<target>
+)";
+
+std::string TargetDescription::to_xml() const noexcept {
+  FeatureStream fs{};
```
Comment (rocallahan) [inline src/TargetDescription.cc:119]: Don't need `{}`

File: src/TargetDescription.cc [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+    target_features.push_back(TargetFeature::AVX);
+  }
+
+  if (cpu_features & rr::GdbServerConnection::CPU_PKU) {
+    DEBUG_ASSERT((arch == rr::x86 || arch == rr::x86_64) && "unexpected arch");
+    target_features.push_back(TargetFeature::PKeys);
+  }
+}
+
+static constexpr auto Header = R"(<?xml version="1.0"?>
```
Comment (rocallahan) [inline src/TargetDescription.cc:113]: Don't need to make this `constexpr` or `auto`. Also, name should be `header` or possibly `HEADER`. E.g. use
`static const char header[] = ...`

Comment (rocallahan) [reply src/TargetDescription.cc:113]: ... or else just write it inline, that's fine

File: src/TargetDescription.h [hunk 14]
```diff
@@ -0,0 +1,27 @@
+#pragma once
+#include "kernel_abi.h"
+#include <cstdint>
+
+namespace rr {
+
+struct GdbServerRegisterValue;
+
+using u32 = std::uint32_t;
```
Comment (rocallahan) [inline src/TargetDescription.h:9]: Please don't do this. Just use `uint32_t` everywhere (no `std::`)

**Review by rocallahan — CHANGES_REQUESTED** [review]

**Comment by rocallahan** [comment]
Also, it looks like tests are failing in CI on x86-64?

File: src/TargetDescription.cc (hunk 13 above)
Comment (theIDinside) [reply src/TargetDescription.cc]: Changed it to `header`, but refrained from inlining it directly as raw string literals make code look weird, though that's a personal taste though, and I'll gladly change if it you like.

**Comment by theIDinside** [comment]
> Also, it looks like tests are failing in CI on x86-64?

Where can I see these tests? It says "All checks have passed" here.

File: src/TargetDescription.h [hunk 15]
```diff
@@ -0,0 +1,27 @@
+#pragma once
```
Comment (rocallahan) [inline src/TargetDescription.h:1]: We don't use this. Use #define guards instead like our other .h files.

Comment (rocallahan) [reply src/TargetDescription.h:1]: Also, put
```
/* -*- Mode: C++; tab-width: 8; c-basic-offset: 2; indent-tabs-mode: nil; -*- */
```
on the first line like our other source files.

File: src/TargetDescription.cc [hunk 16]
```diff
@@ -0,0 +1,128 @@
+#include "TargetDescription.h"
```
Comment (rocallahan) [inline src/TargetDescription.cc:1]: Put
```
/* -*- Mode: C++; tab-width: 8; c-basic-offset: 2; indent-tabs-mode: nil; -*- */
```
on the first line like our other .cc files.

File: src/TargetDescription.cc [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+#include "GdbServerConnection.h"
+#include "kernel_abi.h"
+#include <sstream>
```
Comment (rocallahan) [inline src/TargetDescription.cc:4]: Order #includes like our other files: first `TargetDescription.h`, then the C++ library headers, then the rr headers.

File: src/TargetDescription.h [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+#include "kernel_abi.h"
+#include <cstdint>
```
Comment (rocallahan) [inline src/TargetDescription.h:3]: As above, reorder these includes.

File: src/TargetDescription.h [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace rr {
+
+struct GdbServerRegisterValue;
```
Comment (rocallahan) [inline src/TargetDescription.h:7]: I think you can remove this since it's not used.

File: src/TargetDescription.h [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+  SSE,
+  Linux,
+  Segment,
+  AVX,
+  PKeys,
+  FPU,
+};
+
+class TargetDescription {
+  SupportedArch arch;
```
Comment (rocallahan) [inline src/TargetDescription.h:20]: Put this below the public members in a `private:` section.

File: src/TargetDescription.cc [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+
+namespace rr {
+
+class FeatureStream {
+public:
+  std::string result() { return stream.str(); }
```
Comment (rocallahan) [inline src/TargetDescription.cc:10]: Add `using namespace std;` after the includes and don't use the `std::` prefix (except for `std::move` where clang requires it...)

**Review by rocallahan — CHANGES_REQUESTED** [review]

**Comment by rocallahan** [comment]
> Where can I see these tests? It says "All checks have passed" here.

Yep, that looks good!

**Comment by GitMensch** [comment]
@theIDinside as far as I understood the review that's nearly finished and only needs minor tweaks.

**Comment by theIDinside** [comment]
> @theIDinside as far as I understood the review that's nearly finished and only needs minor tweaks.
> 
> It would be nice if this could be finished as it is specified as a dependency of #3776 which (after rebase including these changes) is a real nice addition.

I'm on it

**Comment by rocallahan** [comment]
I fixed the final nits, squshed and merged: f7f4c298f8acac3c467444471a2b359495c83348

Thanks!!!

**Comment by GitMensch** [comment]
currently this is in [a separate branch](https://github.com/rr-debugger/rr/tree/dynamic-target-descriptions) only... the referenced #3776 needs to be rebased on this (and likely still target the master)

**Comment by rocallahan** [comment]
Oops I pushed to the wrong branch. Fixed.

**Pull request closed without merging**